util/wrappergen/wrappergen.py

```diff
@@ -305,6 +308,43 @@ def main():
     else:
         print("Skipping accelerator generation!")
 
+    # Generate xdma for the whole cluster
+    snax_xdma_cfg = None
+    for i in range(num_cores):
+        if "snax_xdma_cfg" in cfg_cores[i]:
+            snax_xdma_cfg = cfg_cores[i]["snax_xdma_cfg"]
+    if (snax_xdma_cfg is not None):
+        tpl_rtl_wrapper_file = args.tpl_path + "snax_xdma_wrapper.sv.tpl"
+
+        tpl_rtl_wrapper = get_template(tpl_rtl_wrapper_file)
+
+        gen_file(
+            cfg=cfg["cluster"],
+            tpl=tpl_rtl_wrapper,
+            target_path=args.gen_path + cfg["cluster"]["name"] + "_xdma/",
+            file_name=cfg["cluster"]["name"] + "_xdma_wrapper.sv",
+        )
+
+        print(args.gen_path)
+        gen_chisel_file(
+            chisel_path=args.chisel_path,
+            chisel_param="snax.xdma.xdmaTop.xdmaTopGen",
+            gen_path=" --clusterName " + str(cfg["cluster"]["name"]) +
+            " --tcdmDataWidth " + str(cfg["cluster"]["data_width"]) +
+            " --axiDataWidth " + str(cfg["cluster"]["dma_data_width"]) +
+            " --addressWidth " + str(cfg["cluster"]["addr_width"]) +
+            " --tcdmSize " + str(cfg["cluster"]["tcdm"]["size"]) +
+            " --readerDimension " + str(snax_xdma_cfg["reader_agu_dimension"]) +
+            " --writerDimension " + str(snax_xdma_cfg["writer_agu_dimension"]) +
+            " --readerBufferDepth " + str(snax_xdma_cfg["reader_buffer"]) +
+            " --writerBufferDepth " + str(snax_xdma_cfg["writer_buffer"]) +
+            (" --HasMemset " if snax_xdma_cfg["has_memset"] else "") +
+            (" --HasMaxPool " if snax_xdma_cfg["has_maxpool"] else "") +
+            (" --HasTransposer " if snax_xdma_cfg["has_transposer"] else "") +
+            " --target-dir " + args.gen_path +
+            cfg["cluster"]["name"] + "_xdma/"
+        )
```

Comment:
This is smart too because rather than having a scala parameter generated, it can be this one. However, this implies it is very custom to the xDMA only. Not something that is quite "generalized" like the accelerators. I guess it is fine for now since the intent is to have the xDMA a SNAX thing.  But what about the different extensions? That implies that people have to (1) change this wrapper gen to add another extension, (2) modify the scala files for the extension, and (3) add the configurations too.

The SNAX accelerators just need the configuration files since they are "generalized" already and do not need too many modifications. Even the accelerator wrapper names are all centralized in the configuration files.

Reply:
Indeed. it needs to be optimized. It should be done in a seperate PR, I think it is better to rewrite gen_chisel_file function.